assessments/projects/knative/self-assessment.md

```diff
@@ -0,0 +1,577 @@
+    - Key Features :
+        - Ensures ordered, once-only delivery of events.
+        - Can be backed by different messaging systems like Kafka, NATS, etc.
+
+
+9. Creating a Function
+    - Purpose : Allows developers to create a serverless function from a simple piece of code without worrying about the underlying infrastructure.
+    - Common Operations :
+        - Develop Function : Write the function code in a supported language.
+        - Define Function Configuration : Specify the function's attributes, dependencies, and triggers in a configuration file.
+        - Deploy Function : Use a command like "kn func deploy" to deploy the function to a Knative environment.
+
+
+10. Updating a Function
+    - Purpose : Modify existing serverless functions to adapt to new requirements or fix issues.
+    - Common Operations :
+        - Update Code : Modify the function code as needed.
+        - Update Configuration : Adjust the function's configuration file to reflect any changes in dependencies, environment variables, etc.
+        - Redeploy Function : Redeploy the function using the deployment command, which often automatically builds and updates the function in the Knative environment.
+
+
+11. Deleting a Function
+    - Purpose : Remove a function from the Knative environment when it is no longer needed.
+    - Common Operations :
+        - Delete Command : Execute a command like "kn func delete" to remove the function from the environment.
+
+
+12. Monitoring and Logging
+    - Purpose : Provide insights into the function's performance and issues.
+    - Common Operations :
+        - Access Logs : Retrieve logs generated by the function for debugging and monitoring purposes.
+        - Performance Metrics : Utilize integrated monitoring tools to track various metrics such as invocation count, execution time, and resource usage.
+
+
+13. Scaling
+    - Purpose : Automatically scale function instances based on demand.
+    - Common Operations :
+        - Auto-Scaling : Functions are automatically scaled up or down based on the number of incoming requests or events, leveraging Knative's built-in scaling capabilities.
+        - Manual Scaling : Optionally, developers can specify scaling parameters, like min and max number of instances, in the function configuration.
+
+
+14. Event Binding
+    - Purpose : Connect functions to various event sources for event-driven architectures.
+    - Common Operations :
+        - Define Event Sources : Specify event sources in the function configuration.
+        - Event Processing : Write function logic to process incoming events from the defined sources.
+
+
+15. Versioning and Revisions
+    - Purpose : Manage different versions of a function.
+    - Common Operations :
+        - Create Revisions : Each update to a function can create a new immutable revision.
+        - Rollback : Ability to roll back to a previous version of the function if needed.
+
+
+### Goals
```

Comment:
This section intends to cover key goals (typically 3-4) of the project and is not limited to security goals only. Please elaborate on the goals of the project

Reply:
* Minimize the amount of application developer overhead to build and deploy stateless applications
* Offer a streamlined developer experience for application developers 
* Provide an implementation of serverless compute and event handling infrastructure on Kubernetes